### PR TITLE
#11838: Move work_split

### LINKED ISSUE #11838: Move `work_split`
Tracking issue : #11688 , Master Issue : https://github.com/tenstorrent/tt-metal/issues/11513
PR : https://github.com/tenstorrent/tt-metal/pull/11834

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
Issue : #11838, Tracking Issue : #11688, Master Issue : #11513

**Work Done :** 

- Moved file `ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/work_split.hpp` to `tt_metal/common/work_split.hpp` under `tt::tt_metal` namespace
- Updated Bindings : moved `ttnn/cpp/ttnn/deprecated/tt_lib/csrc/tt_lib_bindings_tensor.cpp` to `ttnn/cpp/pybind11/operations/core.hpp` under `ttnn` namespace
- CI : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/10671196837)

CI : 

- [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/10680874044) - PASSED
-  [[post-commit] C++ tests](https://github.com/tenstorrent/tt-metal/actions/runs/10680875449) - PASSED
- [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/10680876960) - PASSED
-  [(Single-card) Model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/10680881917) - PASSED
-  [(T3K) T3000 model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/10680883157) - PASSED
-  [(T3K) T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/10680884420) - PASSED
-  [(T3K) T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/10680885689) - PASSED
-  [Nightly fast dispatch tests](https://github.com/tenstorrent/tt-metal/actions/runs/10680887037) - Passed as in main
-  [(T3K) T3000 frequent tests](https://github.com/tenstorrent/tt-metal/actions/runs/10680888476) - PASSED
-  [(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/10680889540) - Passed as in main
- [(T3K) T3000 profiler tests](https://github.com/tenstorrent/tt-metal/actions/runs/10680890627) - PASSED
- [metal - Run microbenchmarks](https://github.com/tenstorrent/tt-metal/actions/runs/10739777470) - PASSED

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/multi_core/softmax_op_multi_core.cpp [hunk 1]
```diff
@@ -347,7 +347,7 @@ operation::ProgramWithCallbacks scale_mask_softmax_multi_core(
         uint32_t NCHt = NC*Ht;
         uint32_t num_tile_rows = NC * Ht;
         auto all_device_cores = CoreRange({0, 0}, {grid_size.x - 1, grid_size.y - 1});
-        auto [num_cores, all_cores, core_group_1, core_group_2, num_tile_rows_per_core_group_1, num_tile_rows_per_core_group_2] = split_work_to_cores(grid_size, num_tile_rows, true);
+        auto [num_cores, all_cores, core_group_1, core_group_2, num_tile_rows_per_core_group_1, num_tile_rows_per_core_group_2] = ttnn::operations::core::work_split::split_work_to_cores(grid_size, num_tile_rows, true);
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/normalization/softmax/device/multi_core/softmax_op_multi_core.cpp:351]: We (@arakhmati  and I) tried to keep the python api to mirror the C++ api.   So if we have ttnn.split_work_to_cores we want to have ttnn::split_work_to_cores.   It's ok to have split_work_to_cores in the namespace ttnn::operations::core::work_split but we can allow it to exist in just ttnn namespace as well.

Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/multi_core/softmax_op_multi_core.cpp:351]: Will Update

**Review by eyonland — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

**Review by jdesousa-TT — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

**Review by jdesousa-TT — APPROVED** [review]

**Comment by ayerofieiev-tt** [comment]
Screenshot 2024-09-03 at 9 23 23 AM

😅

**Review by johanna-rock-tt — APPROVED** [review]

**Review by ayerofieiev-tt — DISMISSED** [review]
We seem to be moving work_split.hpp to metal/common.hpp for a wrong reason.

It is not used anywhere in metal, except in 2 files in tests/metal:
* as a comment here `tests/tt_metal/tt_metal/perf_microbenchmark/common/work_split.hpp`. Btw, is this a different file?
* and in this test `tests/tt_metal/tt_metal/perf_microbenchmark/1_compute_mm/test_compute_mm.cpp` which actively refers to TT-NN and benchmarks the compute performance of matmul

I don't reasons to move this file to metal, please correct me if I am missing something.

@eyonland @tt-rkim @VirdhatchaniKN .

**Review by sjameelTT — APPROVED** [review]

**Review by djordje-tt — APPROVED** [review]

**Review by tt-aho — APPROVED** [review]

**Comment by eyonland** [comment]
> * work_split.hpp

Yeah, it is perfectly reasonable to write code to help test your code base.    This may need to get moved out, but until we have a better handle on where the refactoring in tt-metal we be, I'd prefer to leave it here.

File: ttnn/ttnn/core.py [hunk 2]
```diff
@@ -44,6 +44,36 @@ def is_sharded(tensor) -> bool:
...
+    )
+
+
+def num_cores_to_corerange_set(
+    target_num_cores: int,
+    grid_size: ttnn.CoreCoord,
+    row_wise: bool = False,
+):
+    """
+    Create a CoreRangeSet containing the specified number of cores
```
[the remark this answers is not in the document]
Comment (eyonland) [reply ttnn/ttnn/core.py:68]: We should never duplicate functionality.

**Comment by tt-aho** [comment]
We should get rid of `tests/tt_metal/tt_metal/perf_microbenchmark/common/work_split.hpp` if we move the file to metal. Seems like the file was copy pasted into microbenchmark probably because the original file was in eager folder.

I am okay with the move to tt_metal util folder for now. These seem like generic utils that aren't dependent on ttnn.

**Review by eyonland — APPROVED** [review]
Please remove tests/tt_metal/tt_metal/perf_microbenchmark/common/work_split.hpp in the next following PR.

**Comment by VirdhatchaniKN** [comment]
> remove tests/tt_metal/tt_metal/perf_microbenchmark/common/work_split.hpp in the next following PR.

Sure Eyon ( #12204 )

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/10698012125 - passed

**Review by uaydonat — APPROVED** [review]

**Review by sankarmanoj-tt — APPROVED** [review]

**Comment by KalaivaniMCW** [comment]
Hi @dongjin-na , @razorback3 , 
Could you take a look at this PR ?

**Review by razorback3 — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

File: ttnn/ttnn/core.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yan-zaretskiy) [reply ttnn/ttnn/core.py:68]: This has to be cleaned up.

**Review by SeanNijjar — APPROVED** [review]

File: ttnn/ttnn/core.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (VirdhatchaniKN) [reply ttnn/ttnn/core.py:68]: I'll work on this.

**Review by cfjchu — APPROVED** [review]

**Review by yan-zaretskiy — APPROVED** [review]

**Comment by VirdhatchaniKN** [comment]
- https://github.com/tenstorrent/tt-metal/actions/runs/10737900385 - passed
- https://github.com/tenstorrent/tt-metal/actions/runs/10737901357 - passed
- https://github.com/tenstorrent/tt-metal/actions/runs/10737902633 - passed
- https://github.com/tenstorrent/tt-metal/actions/runs/10737905160 - passed
- https://github.com/tenstorrent/tt-metal/actions/runs/10737910023 - passed
- https://github.com/tenstorrent/tt-metal/actions/runs/10739777470 - passed

**Comment by VirdhatchaniKN** [comment]
- https://github.com/tenstorrent/tt-metal/actions/runs/10748299922 - passed
- https://github.com/tenstorrent/tt-metal/actions/runs/10748363120 - passed as in main
- https://github.com/tenstorrent/tt-metal/actions/runs/10748389996 - passed as in main
- https://github.com/tenstorrent/tt-metal/actions/runs/10748403563 - passed

**Pull request merged**